### PR TITLE
RenderComponent - new component to replace ModelComponent

### PR DESCRIPTION
**Pull request by mvaligursky**
Superseeds #2345

Related to #2063

Engine examples are in separate PR: https://github.com/playcanvas/engine/pull/2595

Implementation of new RenderComponent, which is functionally similar to ModelComponent, and main reason for its development is to expose Entity hierarchy of GLB files to engine and eventually Editor. Glb file will be loaded as a hierarchy of entities, and RenderComponents would be added on any of these Entities to represent renderable meshes.

This is implemented as @private functionality for the moment, and the API is subject to change.

New API:
- ContainerResource.instantiateModelEntity - creates an entity and adds ModelComponent for model stores as ContainerResource.model.  

- ContainerResource.instantiateRenderEntity - this is the new functionality, and it builds hierarchy of Entities and adds RenderComponents for relevant entities representing meshes assigned to them.

RenderComponent API (similar to ModelComponent API)
- **.meshInstances** - array of meshesInstances this component renders
- **.hide, .show** - allows attached instances to change visibility
- **.generateWireframe** - generates index buffer used to wireframe rendering
- **.castShadows**
- **.receiveShadows**
- **.castShadowsLightmap**
- **.lightmapped, .lightmapSizeMultiplier**
- **.isStatic**
- **.batchGroupId** Assign meshes to a specific batch group
- **.material** The material that will be used to render the meshes (not used on models of type 'asset').

Example:

```
entity.addComponent('render', {
                type: "sphere",
                material: someMaterial,
                castShadows: false,
                castShadowsLightmap: true,
                lightmapped: true
            });
```
 
Other changes:
- Anim system supports Entity hierarchies and RenderComponents (for animation of Entity transforms, and RenderComponent morph weights)
- Batching - supports batching of meshes rendered using RenderComponent
- lightmapping - supports lightmapping of meshes rendered using RenderComponent

### CODE REVIEW

File: src/framework/components/render/component.js
Comment (slimbuck) [inline src/framework/components/render/component.js:348]: these should probably be concatenated with + instead of , ?

File: src/framework/components/render/component.js
Comment (slimbuck) [inline src/framework/components/render/component.js:801]: might have been useful to support multiple assets given that `Entity` only supports a single render component.

File: src/scene/model.js
Comment (slimbuck) [inline src/scene/model.js:173]: might also want to set `this.meshInstances = []`

File: src/scene/render.js
Comment (slimbuck) [inline src/scene/render.js:11]: is this missing generateWireframe?

File: src/framework/components/render/component.js
Comment (mvaligursky) [reply src/framework/components/render/component.js:120]: it's like an event when root bone has been found.
I might change it to 'onRootBoneChange'

File: src/framework/components/render/component.js
Comment (mvaligursky) [reply src/framework/components/render/component.js]: done, thanks

File: src/framework/components/render/component.js
Comment (mvaligursky) [reply src/framework/components/render/component.js:801]: but an asset represents an array of meshInstances/materials .. so that's probably enough?

File: src/scene/model.js
Comment (mvaligursky) [reply src/scene/model.js:173]: yep, done

File: src/scene/render.js
Comment (mvaligursky) [reply src/scene/render.js:11]: this is function called by model component
render component implement this directly in the component

File: examples/graphics/batching-dynamic.html
Comment (mvaligursky) [reply examples/graphics/batching-dynamic.html:65]: testing everything I can :)

File: src/framework/components/render/component.js
Comment (slimbuck) [reply src/framework/components/render/component.js:801]: I'm thinking of the case where a user wants to add a cigar (render component) to a hand node in the hierarchy, when the node already has a render component attached. User would have to create a child node and attach the cigar there, which actually is probably fine.

Comment (mvaligursky) [reply src/framework/components/render/component.js:801]: yep that's how I'd expect them to do it .. either that, or dont smoke.

File: src/framework/components/render/component.js
Comment (mvaligursky) [reply src/framework/components/render/component.js]: renamed to _onRootBoneChanged

File: src/scene/procedural.js
Comment (willeastcott) [inline src/scene/procedural.js:1063]:
```suggestion
// returns Primitive data, used by ModelComponent and RenderComponent
```

File: src/scene/procedural.js
Comment (willeastcott) [inline src/scene/procedural.js:39]: What happens here if two instances of the engine running on the same page start creating primitives? Such as with [this app](http://playcanvas.github.io/#misc/multi-application.html).

File: src/resources/render.js
Comment (willeastcott) [inline src/resources/render.js:7]: Should this be `pc.Render` resources?

File: src/scene/procedural.js
Comment (willeastcott) [reply src/scene/procedural.js:39]: Oh actually, I see you pass in the graphics device so it's fine. Ignore. 😄

**Review by willeastcott — APPROVED** [review]

**Comment by willeastcott** [comment]
Looks good to me. I've approved. But because this is a major change - please only merge this when you have at least to more approvals.

File: src/resources/render.js
Comment (mvaligursky) [reply src/resources/render.js]: fixed, thanks!

**Review by slimbuck — APPROVED** [review]

**Review by ellthompson — APPROVED** [review]

**Pull request merged**
